subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,574 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_cbor
+#define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
+
+#include <logging/log.h>
+#include <sys/util.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include "cbor_common.h"
+#include "cbor_decode.h"
+#include "cbor_encode.h"
+#include "lwm2m_object.h"
+#include "lwm2m_rw_cbor.h"
+#include "lwm2m_engine.h"
+#include "lwm2m_util.h"
+
+#define CPKT_CBOR_W_SZ(pos, cpkt) \
+	((size_t)(pos) - (size_t)(cpkt)->data - (size_t)(cpkt)->offset)
+
+#define ICTX_CBOR_R_SZ(pos, ictx) \
+		((size_t)pos - (size_t)(ictx)->in_cpkt->data - (ictx)->offset)
+
+enum  cbor_tag { /* www.iana.org/assignments/cbor-tags/cbor-tags.xhtml */
+	CBOR_TAG_TIME_TSTR =    0, /* text string	Standard date/time string */
+	CBOR_TAG_TIME_NUM =     1, /* integer or float	Epoch-based date/time */
+	CBOR_TAG_UBIGNUM_BSTR = 2, /* byte string	Unsigned bignum */
+	CBOR_TAG_BIGNUM_BSTR =  3, /* byte string	Negative bignum */
+	CBOR_TAG_DECFRAC_ARR =  4, /* array		Decimal fraction */
+	CBOR_TAG_BIGFLOAT_ARR = 5, /* array		Bigfloat */
+	CBOR_TAG_2BASE64URL =  21, /* (any)		Expected conversion to base64url encoding */
+	CBOR_TAG_2BASE64 =     22, /* (any)		Expected conversion to base64 encoding */
+	CBOR_TAG_2BASE16 =     23, /* (any)		Expected conversion to base16 encoding */
+	CBOR_TAG_BSTR =        24, /* byte string	Encoded CBOR data item */
+	CBOR_TAG_URI_TSTR =    32, /* text string	URI */
+	CBOR_TAG_BASE64URL_TSTR = 33, /* text string	base64url */
+	CBOR_TAG_BASE64_TSTR =    34, /* text string	base64 */
+	CBOR_TAG_MIME_TSTR =      36, /* text string	MIME message */
+	CBOR_TAG_CBOR =        55799, /* (any)	Self-described CBOR */
+};
+
+static size_t put_time(struct lwm2m_output_context *out,
+		      struct lwm2m_obj_path *path, int64_t value)
+{
+	/* CBOR time output format is unspecified but SenML CBOR uses string format.
+	 * Let's stick into the same format with plain CBOR
+	 */
+	struct tm pTime;
+
+	if (gmtime_r((time_t *)&value, &pTime) != &pTime) {
+		LOG_WRN("unable to convert from secs since Epoch to a date/time construct");
+		return -EINVAL;
+	}
+
+	/* Time construct to a string. Time in UTC, offset to local time not known */
+	char time_str[sizeof("1970-01-01T00:00:00-00:00")] = {0};
```

Comment:
Not a big deal for me, but sometimes upstream reviewers request to define variables at the beginning of the block, could fix that rightaway.